### PR TITLE
[TOPI] Basic x86 schedules

### PR DESCRIPTION
**Pull request by masahi**
Contains the following:
* NCHW conv schedule
* NHWC conv compute, schedule, test case
* multi threading for elemwise/injective ops,  pad, pool, softmax
* multi threaded + parallelized bias + batch norm + relu, fused into conv if possible

cc @yidawang @yzhliu

### CODE REVIEW

File: topi/python/topi/x86/conv2d.py [hunk 1]
```diff
@@ -28,10 +34,68 @@ def traverse(op):
...
-            s[C].parallel(c)
-            s[C].pragma(n, "parallel_launch_point")
+            rc, ry, rx = C.op.reduce_axis
+            fused = s[C].fuse(n, c)
+            s[C].parallel(fused)
+            wo, wi = s[C].split(w, factor=16)
+            s[C].reorder(fused, rc, h, wo, ry, rx, wi) # move rc to outer loop
+            s[C].unroll(rx)
+            s[C].unroll(ry)
+            s[C].vectorize(wi)
```
Comment (yzhliu) [inline topi/python/topi/x86/conv2d.py]: I tested this schedule on AWS C5 instance (skylake-avx512 4-cpu) which does not perform well, about `13 ms` for conv only.
My input is `(1, 64, 56, 56)` and kernel size `(64, 64, 3, 3)`

File: topi/python/topi/x86/conv2d.py [hunk 2]
```diff
@@ -28,10 +34,68 @@ def traverse(op):
...
 
     traverse(outs[0].op)
     return s
+
+
+@generic.schedule_conv2d_nhwc.register(["cpu"])
+def schedule_conv2d_nhwc(outs):
```
Comment (yzhliu) [inline topi/python/topi/x86/conv2d.py]: On the same C5 instance, this one works pretty good, `1.8 ms` for `(1, 56, 56, 64)` input and `(3, 3, 64, 64)` kernel size.

File: topi/python/topi/x86/injective.py [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+    sch: Schedule
+        The computation schedule for the op.
+    """
+    outs = [outs] if isinstance(outs, tvm.tensor.Tensor) else outs
+    x = outs[0]
+    s = tvm.create_schedule([x.op for x in outs])
+    tvm.schedule.AutoInlineInjective(s)
+    if len(s[x].op.axis) == 4:
+        n, c, _, _ = s[x].op.axis
+        fused = s[x].fuse(n, c) # for nhwc layout, fuse n and h
```
Comment (yzhliu) [inline topi/python/topi/x86/injective.py]: I'm still confused with this. why not put it in `schedule_conv2d` ?
and the comment does not match.

File: topi/python/topi/x86/conv2d.py [hunk 4]
```diff
@@ -15,6 +15,12 @@ def traverse(op):
         if tag.is_broadcast(op.tag):
             if op not in s.outputs:
                 s[op].compute_inline()
+            else: # inject custom schedule
+                if len(op.axis) == 4: # schedule bias + bn + relu
+                    n, c, h, w = op.axis
+                    fused = s[op].fuse(n, c)
+                    s[op].parallel(fused)
+                    s[op].vectorize(w)
```
[the remark this answers is not in the document]
Comment (masahi) [reply topi/python/topi/x86/conv2d.py]: this is for parallelizing and vectorizing bias + batchnorm + relu.
To understand this, you need to compile your network (resnet18, say) with nnvm and dump lowered IR after operator fusion.
See this https://github.com/dmlc/nnvm/pull/292

Also see this
https://github.com/dmlc/nnvm/issues/275

File: topi/python/topi/x86/conv2d.py (hunk 1 above)
Comment (masahi) [reply topi/python/topi/x86/conv2d.py]: Looks like the final s[C].vectorize(wi) is not working for this input size (56).
For my use cases, all of input width and height is power of two. I wrote my schedules assuming such input.

File: topi/python/topi/x86/injective.py (hunk 3 above)
Comment (masahi) [reply topi/python/topi/x86/injective.py]: this schedule is not about conv2d. This schedule is used for multi threading injective ops such as pooling, upsampling, and softmax. All elemwise/broadcast ops are also multithreaded with this schedule.

**Comment by masahi** [comment]
@yzhliu this PR is not just about x86 conv schedules, but also contains some change I made to run a whole network run faster in conjunction with NNVM.

I should also mention that my schedule is optimized for my use case (input size being power of two, all conv are 3x3, etc..). Nonetheless, it should be better than the current x86 schedule (which is mostly empty). 
The idea is @yzhliu or some others from community will improve on it, to make it faster.

**Comment by tqchen** [comment]
I am going to merge this given the solution is better than current x86 ones. We can followup with updates to improve it
